Support QGIS Server logs to stderr

elemoine opened the pull request:
## Description

QGIS Server can be configured to write its logs to a file on the file system. This is done through the QGIS_SERVER_LOG_FILE environment variable.

But there is currently no way to make QGIS Server write its logs to stderr. And logging to stderr is desirable in some cases, when executing QGIS Server in a Docker container for example.

This PR adds that capability. With this PR, when QGIS_SERVER_LOG_FILE is set  to the special value "stderr",  QGIS Server will write its logs to stderr.

To avoid duplicating code the `QgsServerLogger` class now inherits from `QgsMessageLogConsole`, which already existed in the QGIS code base before this patch. In this way changes to `QgsMessageLogConsole` will directly benefit the server, with no other changes required to the server code.

Note: I have a [doc patch](https://github.com/qgis/QGIS-Documentation/compare/master...elemoine:ele_logging) ready if this one gets accepted.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [X] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [X] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

pblottiere commented on src/core/qgsmessagelog.h:
```diff
@@ -140,6 +140,9 @@ class CORE_EXPORT QgsMessageLogConsole : public QObject
     QgsMessageLogConsole();
 
+  protected:
+    QString formatLogMessage( const QString &message, const QString &tag, Qgis::MessageLevel level );
```
I think that `formatLogMessage` should be a const method.

elemoine replied:
Yep, will fix it.

pblottiere commented on src/server/qgsserverlogger.cpp:
```diff
@@ -36,44 +36,42 @@ QgsServerLogger *QgsServerLogger::instance()
+    mTextStream << formattedMessage;
+    mTextStream.flush();
+  }
+  else if ( mLogFile.fileName() == "stderr" )
```
What do you think about using `QString::compare()` with the `Qt::CaseInsensitive` flag?

elemoine replied:
Well I don't know. "stderr" being a special value, I would be tempted to just support "stderr". But I don't care too much. If you think it's better to support all the combinations that's ok to me too. You decide :)

pblottiere replied:
Actually, I'm not sure of that either... But, for example, I think that 'STDERR' is used sometimes on Windows. 

So I don't have a strong opinion, but I think it would be a more robust approach :).

elemoine replied:
I decided that you were right and went for the more robust approach :)

pblottiere commented on src/server/qgsserverlogger.cpp:
```diff
@@ -36,44 +36,42 @@ QgsServerLogger *QgsServerLogger::instance()
-{
-  Q_UNUSED( tag );
-  if ( !mLogFile.isOpen() || mLogLevel > level )
+  if ( (! f.isEmpty()) && f != "stderr" && mLogFile.open ( QIODevice::Append ) )
```
Same thing about `QString::compare()`.

pblottiere commented:
Hi @elemoine,

Thanks for this patch, it's much better this way!

> To avoid duplicating code the QgsServerLogger class now inherits from QgsMessageLogConsole

Good idea :)!

> Note: I have a doc patch ready if this one gets accepted.

It seems that a `Q` is missing in your documentation:

```
When setting ``GIS_SERVER_LOG_FILE``...
```

elemoine commented:
> It seems that a Q is missing in your documentation:

Thanks for the sharp eyes! This is fixed.

elpaso commented:
Sorry if I jump in late, but I would be in favor of logging **only** to `stderr` and drop file logging completely, this would be much more in line with what the most common server application do.

Redirect from stderr to whatever other stream is usually just a configuration line away in the web server config.

elemoine commented:
Hi @elpaso! Two comments: 1) this patch is about adding support for logging to `stderr`, not removing support for logging to a file. So if we really want to remove file logging support I'd rather do that with a separate PR. 2) Personally I wouldn't remove file logging, as it's very common, it often makes sense, and removing it would be a backward-incompatible change for users relying on it. I don't see the benefit of this removal.

elpaso commented:
Yeah, mine was more of a proposal for the future.

I do not insist of course, but I'm still convinced that file logging can (and should) be configured at the web server level.

See https://12factor.net/ for how a modern web app should ideally behave.

If you stick to web app conventions you get a ton of robust and out-of-the box goodness for free (log rotation and remote logging with syslog for example).

Logging to a file implemented in the application has always looked wierd to me.

elpaso commented on src/core/qgsmessagelog.cpp:
```diff
@@ -47,12 +49,19 @@ QgsMessageLogConsole::QgsMessageLogConsole()
+QString QgsMessageLogConsole::formatLogMessage( const QString &message, const QString &tag, Qgis::MessageLevel level ) const
+{
+  const QString time = QTime::currentTime().toString();
+  const QString levelStr = level == Qgis::Info ? "INFO" :
```
QStringLiteral( "INFO" )

elemoine replied:
Done.

elpaso commented on src/core/qgsmessagelog.cpp:
```diff
@@ -47,12 +49,19 @@ QgsMessageLogConsole::QgsMessageLogConsole()
+  const QString time = QTime::currentTime().toString();
+  const QString levelStr = level == Qgis::Info ? "INFO" :
+                           level == Qgis::Warning ? "WARNING" :
+                           "CRITICAL";
```
Same

elemoine replied:
Done.

elpaso commented on src/core/qgsmessagelog.cpp:
```diff
@@ -47,12 +49,19 @@ QgsMessageLogConsole::QgsMessageLogConsole()
+                           level == Qgis::Warning ? "WARNING" :
+                           "CRITICAL";
+  const QString pid = QString::number( QCoreApplication::applicationPid() );
+  return QString( "%1 %2 %3[%4]: %5\n" ).arg( time, levelStr, tag, pid, message );
```
QString -> QStringLiteral

elemoine replied:
Also done.

elpaso commented on src/core/qgsmessagelog.h:
```diff
@@ -140,6 +140,9 @@ class CORE_EXPORT QgsMessageLogConsole : public QObject
     QgsMessageLogConsole();
 
+  protected:
+    QString formatLogMessage( const QString &message, const QString &tag, Qgis::MessageLevel level ) const;
```
Perhaps you can to add a default INFO to Qgis::MessageLevel level

elemoine replied:
Good call. Done.

elpaso commented on src/server/qgsserverlogger.cpp:
```diff
@@ -36,44 +36,44 @@ QgsServerLogger *QgsServerLogger::instance()
+    mTextStream << formattedMessage;
+    mTextStream.flush();
+  }
+  else if ( QString::compare( mLogFile.fileName(),  "stderr", Qt::CaseInsensitive ) == 0 )
```
QStringLiteral( "stderr" )

elemoine replied:
Yep, done.

dmarteau commented:
Hi

I would like to bring attention to the fact that the server in Qgis3 is tailored to be embedded in different framework: then, the server should not coerce in any case the way logging is handled,  it should be let to the embedder choice.
See here https://projects.3liz.org/infra-v3/py-qgis-server/blob/master/pyqgisserver/utils/qgis.py#L121 for an  example where python code just hook to QgsMessageLog for redirecting to the `logging` python module which is a standard way to handle logging in python services.

So as in the present situation, just be sure that setting  QGIS_SERVER_LOG_FILE to empty value should disable any particular logging channel - except for the actual QgsMessageLog behavior.

Thx

pblottiere commented:
> See here https://projects.3liz.org/infra-v3/py-qgis-server/blob/master/pyqgisserver/utils/qgis.py#L121 for an example where python code just hook to QgsMessageLog for redirecting to the logging python module which is a standard way to handle logging in python services.

> So as in the present situation, just be sure that setting QGIS_SERVER_LOG_FILE to empty value should disable any particular logging channel - except for the actual QgsMessageLog behavior.

@dmarteau Indeed. We should definitely add a unit test for this use case, just to be sure that we're not introducing any regressions.

elemoine commented:
> So as in the present situation, just be sure that setting QGIS_SERVER_LOG_FILE to empty value should disable any particular logging channel - except for the actual QgsMessageLog behavior.

Yep. When QGIS_SERVER_LOG_FILE is not set the code behaves as before: no log is written. In my first patch logs were written to stderr when QGIS_SERVER_LOG_FILE was not set, but I changed that based on the discussion you and I had at the FOSS4G conference.

elemoine commented:
@pblottiere @elpaso after more thinking I agree with you. And I fully agree with the process described by @pblottiere in his [comment](https://github.com/qgis/QGIS/pull/7814#issuecomment-419347986). We can definitely work on that (Paul and I) when this PR is merged.

Commit:
[FEATURE][needs-doc] Support QGIS Server logs to stderr

This commit makes it possible to configure QGIS Server to make it
write its logs to stderr.

This is done by setting the QGIS_SERVER_LOG_FILE env var to the
special value "stderr".

Commit:
Reformat the QgsMessageLogConsole class docs

Commit:
Fix the QgsMessageLogConsole docs

The QgsMessageLogConsole now writes log messages to stderr as opposed
to stdout.

Commit:
Add docs for more QgsMessageLogConsole methods

elemoine commented:
> See here https://projects.3liz.org/infra-v3/py-qgis-server/blob/master/pyqgisserver/utils/qgis.py#L121 for an example where python code just hook to QgsMessageLog for redirecting to the logging python module which is a standard way to handle logging in python services.

I wanted to have a look, but we have no access to your GitLab :)

elemoine commented:
I took all the comments into account. I fixed the build on Travis. I added three commits that bring improvements to the QgsMessageLogConsole docs.

Commit:
Make QgsServerLogger testable and test it

elpaso commented:
What about moving the logger configuration to a method that you can call from init and from the test code?

elemoine commented:
> What about moving the logger configuration to a method that you can call from init and from the test code?

Yep, that could work.

I finally decided to test `QgsServerLogger` separately from `QgsServer`. That makes the tests more specific and simpler, while avoiding clobbering `test_qgsserver.py` even more. See https://github.com/qgis/QGIS/pull/7814/commits/b4502200614cc999328948ff77b452fb8cb690dc. I am hopeful (and quite confident actually) that Travis will be happy.

elemoine commented:
OK, the Travis build passes. This PR is complete from my perspective, but happy to make more changes if necessary.

pblottiere commented on python/core/auto_generated/qgsmessagelog.sip.in:
```diff
@@ -107,18 +107,41 @@ class QgsMessageLogConsole : QObject
 
-This class outputs log messages to the standard output. Therefore it might
+This class outputs log messages to the standard error. Therefore it might
 be the right choice for apps without GUI.
```
While you are at it, could you replace `apps` by `applications` please?

pblottiere commented on src/core/qgsmessagelog.h:
```diff
@@ -128,19 +128,42 @@ class CORE_EXPORT QgsMessageLogNotifyBlocker
+     * \param message the message to format
+     * \param tag the tag of the message
+     * \param level the log level of the message
+     */
```
Can you add `\since QGIS 3.4` in the documentation?

pblottiere commented on src/core/qgsmessagelog.h:
```diff
@@ -128,19 +128,42 @@ class CORE_EXPORT QgsMessageLogNotifyBlocker
+  protected:
+
+    /**
+     * Format a log message. Used by child classes.
```
Format**s**

pblottiere commented on src/core/qgsmessagelog.h:
```diff
@@ -128,19 +128,42 @@ class CORE_EXPORT QgsMessageLogNotifyBlocker
   public slots:
+
+    /**
+     * Log a message to stderr.
```
Log**s**

pblottiere commented:
@elemoine Just some minor comments on the documentation, otherwise, I think it's ready to be merged.

Thanks a lot four your work!

Commit:
Improve qgsmessagelog doc strings

pblottiere commented:
After a discussion with @elemoine, we decided to rework this PR before merging it in order to add a new `QGIS_SERVER_LOG_STDERR` variable (instead of introducing a `stderr` value for `QGIS_SERVER_LOG_FILE`). 

So, this PR is not ready to be merged yet finally :).

Commit:
[FEATURE][needs-doc] Introduce QGIS_SERVER_LOG_STDERR and deprecate file logging

pblottiere commented on src/server/qgsserverlogger.h:
```diff
@@ -59,9 +58,15 @@ class QgsServerLogger: public QObject
 
+    /**
+     * Activates logging to stderr.
+     */
```
`\since QGIS 3.4`

elemoine commented:
@pblottiere FYI, I updated my [doc patch](https://github.com/qgis/QGIS-Documentation/compare/master...elemoine:ele_logging) accordingly.

Commit:
Make QgsMessageLogConsole::logMessage virtual

pblottiere commented:
@elemoine I just tested your branch and it works well!

Waiting for Travis now :).

pblottiere commented:
Thanks for your work @elemoine!

elemoine commented:
@elpaso for your information, Paul and I added QGIS_SERVER_LOG_STDERR, in preparation for the removal of QGIS_SERVER_LOG_FILE in QGIS 4.0. See https://github.com/qgis/QGIS-Documentation/pull/3014 for the deprecation of QGIS_SERVER_LOG_FILE in QGIS 3.4. Thanks for the suggestions you provided on that front.
